File: 백준/Gold/2470. 두 용액/두 용액.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>

using namespace std;

int main(){

    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);


    int n; cin >> n;
    vector<long> liquids(n);

    for(int i = 0; i < n ; i++){
        long liquid; cin >> liquid;
        liquids[i] = liquid;
    }

    sort(liquids.begin(),liquids.end());

    int left = 0;
    int right = n-1;

    long minValue = liquids[left] + liquids[right];
    int minLeft = left;
    int minRight = right;
    while(left < right){

        long value = liquids[left] + liquids[right];

        if(abs(value) < abs(minValue)){
            minLeft = left;
            minRight = right;
            minValue = value;
        }

        if(value > 0)
            right--;
        else
            left++;
    }

    cout << liquids[minLeft] << " " << liquids[minRight];
}

File: 백준/Gold/2470. 두 용액/두 용액.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;


public class Main {

	private static long[] liquids;

	private static long[] solution(){

		Arrays.sort(liquids);

		int minValueLeftIndex = 0;
		int minValueRightIndex = liquids.length-1;
		long minValue = liquids[minValueLeftIndex] + liquids[minValueRightIndex];

		int left = 0;
		int right = liquids.length-1;
		while(left < right){

			long value = liquids[left] + liquids[right];
			if(Math.abs(value) < Math.abs(minValue)){
				minValue = value;
				minValueLeftIndex = left;
				minValueRightIndex = right;
			}

			if(value > 0)
				right--;
			else
				left++;
		}

		return new long[]{liquids[minValueLeftIndex],liquids[minValueRightIndex]};
	}

	public static void main(String args[]) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());

		liquids = new long[n];

		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i = 0; i < liquids.length; i++)
			liquids[i] = Long.parseLong(st.nextToken());

		long[] answer = solution();
		System.out.println(answer[0] + " " + answer[1]);
	}
}